### ISSUE TITLE
[LICM] wrongcode at -O2

### ISSUE THREAD

**Issue by JonPsson** [llvm/llvm-project#57780]
Bisected to 0776c48 "Recommit "[LICM] Only create load in ph when promoting load or store doesn't exec."

This program is supposed to print 0:

```console
$ clang -O0 wrong0.i -o a.out ; ./a.out
0
$ clang -O2 wrong0.i -o a.out ; ./a.out
5
```

```c
int printf(const char *, ...);

int a = 0, b = 1;
short c = 0;
long d = 0;
int *e = &b;

int f(int g) { return a / g; }

int main() {
  int h = 10;

i:
  if (h) {
    int j = 0;
    if (d) {
      long k = 4;
    l:
      if (b)
        goto i;
    }
    h = 0;
    for (; h < 6; ++h) {
      if (f(*e))
        ;
      else
        goto l;
      c = 5;
      for (; b; b--)
        ;
    }
  }

  printf("%d\n", c);
}
```

[testcase.tar.gz](https://github.com/llvm/llvm-project/files/9583448/testcase.tar.gz)

**Comment by JonPsson** [comment]
(-target s390x-linux-gnu)

**Comment by nikic** [comment]
Reduced a bit:
```
int printf(const char *, ...);

int a = 0, b = 1;
short c = 0;
long d = 0;

int main() {
  int h = 10;

i:
  if (h) {
    int j = 0;
    if (d) {
      long k = 4;
    l:
      if (b)
        goto i;
    }
    h = 0;
    for (; h < 2; ++h) {
      if (!a)
        goto l;
      c = 5;
    }
  }

  printf("%d\n", c);
}
```
With this it's possible to run through alivecc, which doesn't report any issues :/ So either alive misses the miscompile or the issue is in the backend.

**Comment by JonPsson** [comment]
It's not in the backend - see previous comment about the constant '5' being loaded before isel.

**Comment by nikic** [comment]
Oh yeah, you're right. So alive just misses this.

This is the IR before the relevant LICM run: https://gist.github.com/nikic/f60e97976eab9d73ccdb5d9e12c9b55e

**Comment by nikic** [comment]
A bit further reduced: https://gist.github.com/nikic/d4412c2289144a70168b9bfe9b784308

For some reason, SafetyInfo believes that the store to `@c` is guaranteed to execute.

**Comment by nikic** [comment]
A further reduction:
```
@c = global i16 0, align 2
 
define void @test() {
entry:
  br label %loop

loop:                                             ; preds = %loop.latch, %entry
  %v = phi i32 [ 10, %entry ], [ 0, %loop.latch ]
  %c = icmp eq i32 %v, 0
  br i1 %c, label %loop.exit, label %loop.cont

loop.cont:                                        ; preds = %loop
  br i1 false, label %loop.irreducible, label %loop.latch

loop.irreducible:                                 ; preds = %loop.latch, %loop.cont
  store i16 5, ptr @c, align 2
  br label %loop.latch

loop.latch:                                       ; preds = %loop.irreducible, %loop.cont
  br i1 false, label %loop.irreducible, label %loop

loop.exit:                                        ; preds = %loop
  ret void
}
```

This one has an even more obvious failure mode, because the store gets hoisted into the entry block.

Clearly LoopSafetyInfo is broken in the presence of irreducible control flow :(

**Comment by nikic** [comment]
Candidate patch: https://reviews.llvm.org/D134279
